tfhe/examples/fhe_strings/server_key/pattern/contains.rs

```diff
@@ -0,0 +1,247 @@
+use crate::ciphertext::{FheAsciiChar, FheString, GenericPattern};
+use crate::server_key::pattern::{CharIter, IsMatch};
+use crate::server_key::ServerKey;
+use rayon::prelude::*;
+use rayon::range::Iter;
+use tfhe::integer::BooleanBlock;
+
+impl ServerKey {
+    // Compare pat with str, with pat shifted right (in relation to str) the number given by iter
+    fn compare_shifted(
+        &self,
+        str_pat: (CharIter, CharIter),
+        par_iter: Iter<usize>,
+        ignore_pat_pad: bool,
+    ) -> BooleanBlock {
+        let mut result = self.key.create_trivial_boolean_block(false);
+        let (str, pat) = str_pat;
+
+        let matched: Vec<_> = par_iter
+            .map(|start| {
+                let str_chars = str.clone().skip(start);
+                let pat_chars = pat.clone();
+
+                if ignore_pat_pad {
+                    let str_pat = str_chars.into_iter().zip(pat_chars).par_bridge();
+
+                    self.asciis_eq_ignore_pat_pad(str_pat)
+                } else {
+                    let a: Vec<&FheAsciiChar> = str_chars.collect();
+                    let b: Vec<&FheAsciiChar> = pat_chars.collect();
+
+                    self.asciis_eq(a.into_iter(), b.into_iter())
+                }
+            })
+            .collect();
+
+        for match_case in matched {
+            // One of the possible values of pat must match the str
+            self.key.boolean_bitor_assign(&mut result, &match_case);
+        }
```

Comment:
Actually here, you should be able to transform all these matched booleanblocks into a RadixCiphertext, and use scalar_ne_parallelized(&radix, 0) on it, it should be faster

Reply:
Thanks! Very useful